### ISSUE TITLE
Design a database schema for swaps

### ISSUE THREAD

**Issue by thomaseizinger** [comit-network/comit-rs#2485]
## Problem

We want to store swaps in an SQLite database (the original spike provides reasoning on why we are using SQLite.)
SQLite requires us to define an SQL schema before we can store data.
The scope of this ticket is to define said schema in a structured way.

## Goal

1. Have a database schema for all the data related to swaps:
a) all the data provided by the user on the API
b) all the data learned as part of the communication phase
c) any data we can re-fetch from the blockchain is **out of scope**

2. Provide documentation on the design choices of the schema as a spike document.
Database schemas boil down to degrees of [normalization](https://en.wikipedia.org/wiki/Database_normalization). Perfect normalization is often not desirable because it leads to performance problems due to a large number of JOINs to retrieve related data.

## Schema requirements

1. We don't have all the data available from the beginning, i.e. some of the data is only learned during the communication phase.
a. We need to be able to distinguish between these two kinds of swaps (i.e. "created" and "finalized")

2. We need a reasonably efficient way of determining the "types" of a swap: Given the UUID of a swap, find out, which protocols the swap is composed of and which role the user is taking in this swap.

3. This should support han (Bitcoin + Ethereum), herc20 and halight (Bitcoin). Don't design for Grin but keep in mind that not all swaps have a secret hash. The secret hash and identities is actually linked to a specific protocol.

## Prior work

Originally, we split the design work into two tickets: https://github.com/comit-network/comit-rs/issues/2173 and https://github.com/comit-network/comit-rs/issues/2177.
After some attempts (https://github.com/comit-network/comit-rs/pull/2431, https://github.com/comit-network/comit-rs/pull/2432, https://github.com/comit-network/comit-rs/pull/2433, https://github.com/comit-network/comit-rs/pull/2434, https://github.com/comit-network/comit-rs/pull/2435) of doing those tickets, it became clear that a more structured approach is needed to tackle this problem.

@bonomat did some work in this regard already: https://3.basecamp.com/4403044/buckets/16118249/messages/2583518881

@thomaseizinger wrote down some ideas here: https://blog.eizinger.io/p/4vSOQXyW45

## Recommendation

1. Familiarize yourself with the topic of database normalization (up to 3rd normal form should be enough)
2. Normalize our data according to those rules
3. Document the fully normalized solution
4. Make choices on how to diverge from the fully normalized solution and document those choices.
